dawn/src/dawn/Optimizer/PassMultiStageMerger.cpp

```diff
@@ -0,0 +1,156 @@
+    return ReturnType(std::nullopt, multiStageLoopOrder);
+
+  // Check all possible loop orders if there aren't any vertical conflicts
+  for(auto loopOrder : possibleLoopOrders) {
+    auto conflict = hasVerticalReadBeforeWriteConflict(multiStageDependencyGraph, loopOrder);
+    if(!conflict.CounterLoopOrderConflict)
+      return ReturnType(multiStageDependencyGraph, loopOrder);
+  }
+
+  return ReturnType(std::nullopt, multiStageLoopOrder);
+}
+
+/// @brief Check if we can merge the first multistage into the other, possibly changing the loop
+/// order.
+/// @returns The new dependency graphs of the multi-stage (or NULL) and the new loop order
+template <typename ReturnType =
+              std::pair<std::optional<iir::DependencyGraphAccesses>, iir::LoopOrderKind>>
+ReturnType isMergable(const iir::MultiStage& thisMS, const iir::MultiStage& otherMS) {
+  iir::LoopOrderKind thisLoopOrder = thisMS.getLoopOrder();
+  for(const auto& thisStage : thisMS.getChildren()) {
+    auto dependencyGraphLoopOrderPair = isMergable(*thisStage, thisLoopOrder, otherMS);
+    if(dependencyGraphLoopOrderPair.first) {
+      return dependencyGraphLoopOrderPair;
+    }
+  }
+  return ReturnType(std::nullopt, thisLoopOrder);
+}
+
+PassMultiStageMerger::PassMultiStageMerger(OptimizerContext& context)
+    : Pass(context, "PassMultiStageMerger") {
+  dependencies_.push_back("PassSetStageGraph");
+}
+
+bool PassMultiStageMerger::run(const std::shared_ptr<iir::StencilInstantiation>& instantiation) {
+  const int maxBoundaryExtent = context_.getOptions().MaxHaloPoints;
+  for(const auto& stencil : instantiation->getStencils()) {
+    if(stencil->getChildren().size() < 2)
+      continue;
+
+    auto& metadata = instantiation->getMetaData();
+    std::unique_ptr<iir::Stencil> newStencil = std::make_unique<iir::Stencil>(
+        metadata, stencil->getStencilAttributes(), stencil->getStencilID());
+    newStencil->setStageDependencyGraph(
+        iir::DependencyGraphStage(*stencil->getStageDependencyGraph()));
+
+    // NOTE: The Stages in the same multi-stage are assumed to have no counter loop-order vertical
+    // dependencies so we can treat each multistage independently.
+    for(auto [thisIdx, thisMS] : enumerate(stencil->getChildren())) {
+      int mergeIdx = -1;
+      for(auto [otherIdx, otherMS] : enumerate(newStencil->getChildren())) {
+        // 1) Are the loop orders compatible?
+        if(loopOrdersAreCompatible(thisMS->getLoopOrder(), otherMS->getLoopOrder())) {
+          // 2) Can we merge the stage without violating vertical dependencies?
+          auto dependencyGraphLoopOrderPair = isMergable(*thisMS, *otherMS);
+          auto multiStageDependencyGraph = dependencyGraphLoopOrderPair.first;
+          if(multiStageDependencyGraph &&
+             !multiStageDependencyGraph->exceedsMaxBoundaryPoints(maxBoundaryExtent)) {
+            otherMS->setLoopOrder(dependencyGraphLoopOrderPair.second);
+            mergeIdx = otherIdx;
+          }
```

Comment:
should we check max halo size before merging? Or do you prefer to merge and than split again?

Reply:
Changed `maxBoundaryExtent` to `maxHaloPoints`.